### ISSUE TITLE
!!!!! Iframe being used

### ISSUE THREAD

**Issue by Cis112233** [Khan/live-editor#557]
https://www.khanacademy.org/computer-programming/spin-off-of-how-to-create-an-iframe-real/3009491503
That is a spin off that I made and as you can see that program opens gamestop.com in a iframe. This is bad. This is a big issue. @pamelafox

**Comment by Cis112233** [comment]
@kevinbarabash

**Comment by Gigabyte-Giant** [comment]
Iframe doesn't load for me, as the CSP blocks it.

Tested on:
 - Chrome 48.0.2564.116 (64-bit)
 - Safari 9.0.3 (11601.4.4)
 - Firefox 44.0.2

Also, just a note, everyone who watches this repository gets a notification when you report an issue/comment, so you don't have to ping Pamela and/or Kevin.

**Comment by Cis112233** [comment]
Perhaps this is another Edge issue?
![2016-02-29](https://cloud.githubusercontent.com/assets/9143435/13410922/5e7ab6ba-df08-11e5-8ea1-f1d370059a9c.png)

**Comment by Gigabyte-Giant** [comment]
Interesting, I was about to say something about Edge. Mind providing a screenshot of the console @Cis112233?

**Comment by MikaalSky** [comment]
Edge isn't exactly the best browser for security things, really. It's a lot like internet explorer in that aspect, really.

Makes me seriously worry about how safe my information is from hackers, MS wise.

**Comment by Cis112233** [comment]
I will in one minute. I have several full screens of errors

**Comment by MikaalSky** [comment]
Just post them in text format.

**Comment by VirusKA** [comment]
I got it to work on my Chrome for Mac

**Comment by Cis112233** [comment]
Can you do the errors? @viruska I can't get on the computer that khan works on right now.

**Comment by VirusKA** [comment]
no i didn't get errors....

**Comment by Cis112233** [comment]
Wow. I got a few pages of errors.

**Comment by Cis112233** [comment]
![2016-02-29 1](https://cloud.githubusercontent.com/assets/9143435/13411375/ba411a96-df0a-11e5-96cc-d5f02be32d23.png)
![2016-02-29 2](https://cloud.githubusercontent.com/assets/9143435/13411377/ba5373e4-df0a-11e5-8765-8e96a485aef3.png)
![2016-02-29 3](https://cloud.githubusercontent.com/assets/9143435/13411376/ba43f3c4-df0a-11e5-89e6-f5aacfe5726d.png)
Three pages of error

**Comment by Gigabyte-Giant** [comment]
Chrome, Firefox, and Safari (among other browsers, I'm sure), block the frame from loading if the CSP is violated. Edge does not, as it would appear.

**Comment by Cis112233** [comment]
![capture](https://cloud.githubusercontent.com/assets/9143435/13411483/422ad2ee-df0b-11e5-9a04-193258c40757.PNG)
So on that page of errors you have the 99% https security errors and one warning about iframe towards the bottom

**Comment by Cis112233** [comment]
SEC7131: Security of a sandboxed iframe is potentially compromised by allowing script and same origin access.

**Comment by Gigabyte-Giant** [comment]
I have to agree with what @ShadowC-X-11 mentioned on https://github.com/Khan/live-editor/issues/427#issuecomment-190437248. I seriously doubt that this is a problem with the CSP, but the `sandbox` attribute and MS Edge.

When I get a few free minutes, I'm going to cook up a little test. I'll post what I find, if anything.

**Comment by Cis112233** [comment]
Well this is not just an edge bug. Virus loaded the page up also. Perhaps it is a edge bug and a chrome version <whatever version virus has>

**Comment by VirusKA** [comment]
I'm gonna correct my statement. I meant the original post with khanacademy.org as the link being used. The one linked in this issue I cannot load.

**Comment by ShadowC-X-11** [comment]
Well, since we're talking about loading stuff that shouldn't be loadable, you can also use window.open in the sandbox. There's also a few other things to consider if we want to block off the world of downloads and exiting the sandbox, along with loading data you shouldn't, including:

- The fact that you can download Blobs by doing a few small things, which is a lot easier in Edge/IE due to navigator.msSave[OrOpen]Blob
- You can use a FileReader to open a window with a data URL made from a Blob, which will then allow a download 
- window.open with a javascript: URL can allow a new, no limit environment for extra special stuff (like turning an image into data that can be put in a <canvas> in the host)

The first part, luckily, is blocked in other, older browser. If there was no CSP, an effective replacement would look like:
```
function downloadBlob(b, name){
    var d = document.createElement("a");
    d.download = name; d.href = URL.createObjectURL(b);
    d.click();
}
```
In Chrome (and Firefox I believe), Data URLs and Blob URLs don't work because of the CSP.
Anyways, if these are too severe, you'll need to do some crazy stuff to fix it. That will mostly include a bunch more libraries being broken. I enjoy the HTML environment, however, unless browsers fix these bugs, we're doomed.

**Comment by ShadowC-X-11** [comment]
By the way, I'd also like to bring to light, if not already viewed by others, issue #427. If an IFrame sandbox problem is to blame for this, it's also to blame for the problem shown there. It's very probable that the 2 are linked. If @Gigabyte-Giant proves the theory, the main problem is for Edge, and we just put an issue on Edge Feedback. Otherwise, the whole problem has to do with CSP, which means that Edge has a major problem, and we'll have to temporarily stop using it in case we have somebody use these problems to their advantage.

**Comment by thomasmorelli** [comment]
Yep, it is definitely an Edge issue, Edge seems to be ignoring the `Content Security Policy directive: "child-src *.kastatic.org www.khanacademy.org 'self'"`. Is this a bug with the programming of Edge, or one that can be fixed within KA?

**Comment by ShadowC-X-11** [comment]
@thomasmorelli This is definitely an Edge issue. This needs to be reported to MS so they can fix it.

**Comment by kevinbarabash** [comment]
> Otherwise, the whole problem has to do with CSP, which means that Edge has a major problem, and we'll have to temporarily stop using it in case we have somebody use these problems to their advantage.

@ShadowC-X-11 are you suggesting that we block people using Edge?

**Comment by kevinbarabash** [comment]
I'm going to install an Edge vm and poke around with this a bit.

**Comment by Cis112233** [comment]
@kevinbarabash that would be a problem for a lot of people if you ban Edge. Perhaps just check on `document.createElement `if the document being created is an iframe. Just have it check the args and if it is iframe block it. Would that work?

**Comment by kevinbarabash** [comment]
> that would be a problem for a lot of people if you ban Edge

Agreed.

> Perhaps just check on document.createElement if the document being created is an iframe. Just have it check the args and if it is iframe block it. Would that work?

Good suggestion.  Creating a wrapper should make it more robust... something like:
```
var createElement = HTMLDocument.prototype.createElement;
HTMLDocument.prototype.createElement = function(tagname) {
    if (tagname === 'iframe') {
         throw new Error('creating iframes is not allowed');
    }
    return createElement.call(this, tagname);
};
```

**Comment by kevinbarabash** [comment]
TODO:
- [ ] determine how many people use `iframe`s in the webpage environment
- [ ] double check that we don't have any talk-throughs, etc. that use `iframe`s
- [ ] implement the above wrapper in an IIFE so that people can't access the original `createElement`
- [ ] test that the wrapper works with jQuery (doesn't break it, but doesn't allow the creation of `iframe`s)

**Comment by omridevk** [comment]
hmm

what if somebody really bad and just do something like  ```document.querySelector('iframe').cloneNode(false)```
then modify the iframe attributes and attach it back wherever, have not tested if this actually works but I don't see a reason why it shouldn't.

not even sure if it is relevant here, just glanced by this bug without actually knowing the repo :/ sorry if this is not helpful at all and I misunderstood the issue all together.

**Comment by ShadowC-X-11** [comment]
Alright, I can confirm that on the latest Insider build, this problem is still present. While I've got the information on my mond, also on Edge, you can access UserMedia and throw that in a <video> tag, but other browsers can't. You can also use navigator.msSaveBlob to spam downloads. A tiny bit of a problem for sure.

**Comment by kevinbarabash** [comment]
We should prevent access to APIs that don't have cross browser support which we don't have a polyfill for.

**Comment by Cis112233** [comment]
@ShadowC-X-11 Could you create a program with some of the different bugs so we can block them?

**Comment by kevinbarabash** [comment]
I'd rather have a whitelist.  What things on `navigator` should users have access to?

**Comment by ShadowC-X-11** [comment]
Alright. So it's easier for me to say what should be blocked, because there's a bunch of extra things in `navigator` that's nice. There's also a couple of other APIs that probably could be blocked, so here we go.
By the way, some of these are prefixed on certain browsers, so you might want to take a look at Mozilla's documentation. MS Edge is left out, but Microsoft also has documentation pertaining to IE11 and Edge.
`navigator.getUserMedia/navigator.mediaDevices` - They both allow access to the camera and mic. One returns a `MediaStream` and the other returns a `Promise` that eventually gives a `MediaStream`.
'navigator.geolocation' - This allows the program to know the longitude and latitude of the user, along with how fast they're going.
`navigator.msSave[OrOpen]Blob` - This allows the program to download files to the users system. As the ms prefix suggests, it's not cross platform (it's Edge and IE only), but that can be fixed with polyfills. But due to your CSPs, they wouldn't work either, as most of them depend on using <a> tags with a blob: link generated with `window.createObjectURL`.
'[window.]addEventListener("gamepadconnected", function(gamepad){}), [window.]addEventListener("gamepaddisconnected", function(gamepad){}), navigator.getGamepads()' - This allows access to gamepads connected to the user's computer. This is also prefixed with webkit in certain browsers.

For good measure, while you're blocking the `<iframe>` from use in `document.createElement`, you should also block `<video>`. Also, if the audio tag should be blocked, get rid of that. However, also block `window.Audio` in that case. It constructs an audio tag with the first parameter as the src.
Also, don't use JSLint for this. I already found a way to bypass it in the PJS environment. You can see my example [here](https://www.khanacademy.org/computer-programming/leaving-the-sandbox/5049748660027392). It has a lot of different bugs exploited, so take a look at that if you want to crush some problems in the PJS area while you're at it. (if you get around to it, though, please allow access to Date() in the PJS environment.)
Anyways, this should probably do the trick. If you really want a whilelist for `navigator`, all it really does is allow access to the web browsers data, so might as well block it all, except userAgent

**Comment by Cis112233** [comment]
So while messing around with the iframe glitch... You can imbed videos. If you look I have put a video on the program I sent a link to at the top

**Comment by ShadowC-X-11** [comment]
Seriously? Rickrolling web devs?

**Comment by Cis112233** [comment]
Opps
Seriously? Rickrolling web devs?

---

**Comment by Cis112233** [comment]
We need to get this done

**Comment by kevinbarabash** [comment]
Sorry I haven't done any the TODOs that I outlined above.  I'll spend some time on this next week.
